### DISCUSSION TITLE
Improving typings of `depth` in collections

### DISCUSSION THREAD

**Discussion by tyteen4a03** [payloadcms/payload#6161] [Feature Requests & Ideas]
Right now, Payload types does not have the concept of depth-aware types. For a typical collection, when a relationship type is generated into the schema, it hints `RelatedEntity | number` (or `string` depending on your ID type). This isn't useful behaviour for the consumers as they now need to either cast the variable to `RelatedEntity` to get type completion or do some type guard checks to make sure the variable resolved to `RelatedEntity`.

I'm not sure how this can be improved in TypeScript but just throwing this out there.

**Comment by jmikrut** [comment]
Hey there - a solid fix here is super important for sure, but it's more complicated than depth alone unfortunately. For example, you might have access control that prevents a user from reading a document - or, in MongoDB at least, you might have deleted a document which will then prevent the document from populating successfully. 

Basically - this problem runs a bit deeper than solely being based on depth alone. But I agree, something needs to be done here. We will figure out a solution one way or another. Will bring this forward in our priorities list.

**Comment by mbaroukhcw** [comment]
Hi.

As a dirty fixe, I generate type in a file named `payload-typs-raw.ts` and use this :

```typescript

import type * as PL from './payload-types-raw'

type ElementType<T> = T extends (infer U)[] ? U : never;

type TransformType<T> = {
  [K in keyof T]: K extends "layout" ? TransformType<ElementType<T[K]>>[] : T[K] extends number | infer U ? U : TransformType<T[K]>
  
}
type TransformCollection<T> = {
  [K in keyof T]: TransformType<T[K]>
}

export type Config = {
  globals: PL.Config['globals']
  collections: TransformCollection<PL.Config['collections']>
}

export type Page = TransformType<PL.Page>
<Export all your entities>
````

At least it does the job for me (all my ids are number).

I just suppose that, if there is an ID, the referenced object will always be present.

**Comment by 6TELOIV** [comment]
I recently ran into this with a poly relationship field, and had to ask on stack overflow. jcalz provided a thourough and great answer. I have copied it here for posterity
https://stackoverflow.com/a/79562932/3984972

-----

One problem you're having is that TypeScript does not synthesize [union types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types) during [inference](https://www.typescriptlang.org/docs/handbook/type-inference.html) of [generic](https://www.typescriptlang.org/docs/handbook/2/generics.html) type arguments. When you call a function of the form `function f<T>(arg: A<T>): R<T>`, TypeScript will accept arguments of type `A<X>` for any `X` that's already part of the type.  But if you pass in an argument of a union type like `A<X> | A<Y>`, TypeScript will usually *not* attempt to unify `X` and `Y` into a single type argument `X | Y`. Instead it decides that the call was probably a mistake and issues a warning.  This behavior is intentional and described in [microsoft/TypeScript#23312](https://github.com/Microsoft/TypeScript/issues/23312) and https://stackoverflow.com/q/39905523/2887218.

Of course nothing stops you from *explicitly specifying* the type argument, like this:

```ts
const arr = polyRelationshipMany<TypeA | TypeB | TypeC>(e.items); // okay
/* const arr: {
  relationTo: string;
  value: TypeA | TypeB | TypeC;
}[] */
```

That works, but... the output type isn't really what you want, is it? The resulting array has no idea that `relationTo` can only be one of the known string [literal types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types), or that those string literal types are correlated with the `value`. Instead `TypeA`, `TypeB`, and `TypeC` have been squished together. That means the elements of your array are no longer [discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions):

```ts
e.items.map(x => {
  if (typeof x.value === "number") return 0;
  switch (x.relationTo) {
    case "type-a": return x.value.a; // okay
    case "type-b": return x.value.b; // okay
    case "type-c": return x.value.c; // okay
  }
});

const arr = polyRelationshipMany<TypeA | TypeB | TypeC>(e.items);
arr.map(x => {
  switch (x.relationTo) {
    case "type-a": return x.value.a; // error!
    case "type-b": return x.value.b; // error!
    case "type-c": return x.value.c; // error!
  }
});
```

Ideally you want `arr`'s element type to still be a discriminated union, where each of the union members' `value` property has `number` excluded from it.  This is possible, but it's more complicated to write.  TypeScript won't automatically *distribute* generic function calls over union types in its function arguments, as requested in [microsoft/TypeScript#52295](https://github.com/microsoft/TypeScript/issues/52295), so you'll have to explicitly anticipate this and write for it.

----

Here's one approach:

```ts
function numberGuard<T extends { value: unknown }>(v: T):
  v is ExcludeNumberFromValue<T> {
  return typeof v.value !== "number"
}
function polyRelationshipMany<T extends { value: unknown }>(
  items: T[]
): ExcludeNumberFromValue<T>[] {
  return items.filter(numberGuard);
}
```

where `ExcludeNumberFromValue<T>` is what we have to write. A first attempt could be

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
    Omit<T, "value"> & { value: Exclude<T["value"], number> };
```

using the [`Omit`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) and [`Exclude`](https://www.typescriptlang.org/docs/handbook/utility-types.html#excludeuniontype-excludedmembers) utility types. That does something reasonable for an input type like

```ts
type X = ExcludeNumberFromValue<{ x: Date, value: string | number }>
// type X = Omit<{ x: Date, value: string | number }, "value"> & { value: string }
```

well, it's hard to see what type that is, so we can use something like https://stackoverflow.com/q/57683303/2887218 to show us: 

```ts
type IdX = Id<X>;
// type IdX = { x: Date; value: string; }
```

Hooray! Unfortunately, this does *not* distribute over unions in `T`:

```ts
type Y = Id<ExcludeNumberFromValue<
  { x: Date, value: string | number } | 
  { y: number, value: boolean | number }
>>
// type Y = { value: string | boolean; }
```

We lost `x` and `y` entirely, and squished `value` together. That's just the nature of `Omit` (see https://stackoverflow.com/q/57103834/2887218
).  We need a technique to operate over each union member separately. The go-to approach here is to make it a [distributive conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types). Wrapping it with `T extends unknown ? ⋯ : never` will do it:

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
  T extends unknown ? 
    Omit<T, "value"> & { value: Exclude<T["value"], number> } 
  : never;
```

Now we get the right type for `Y`:

```ts
/* type Y = {
  x: Date;
  value: string;
} | {
  y: number;
  value: boolean;
} */
```

Once we use a conditional type like this it's a little bit easier to use [`infer` within conditional types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types) to pull out the `value` property type, instead of using an [indexed access type](https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html):

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
  T extends { value: infer V } ?
    Omit<T, "value"> & { value: Exclude<V, number> } 
 : never;
```

But that's the same thing.

----

So that type is what we want, mostly... except, we can't directly use it:

```ts
declare function numberGuard<T extends { value: unknown; }>(
  v: T): v is ExcludeNumberFromValue<T> // error!
// A type predicate's type must be assignable to its parameter's type.
```

Oops, TypeScript can't tell that `ExcludeNumberFromValue<T>` is a valid narrowing of `T`.  See https://stackoverflow.com/q/75621955/2887218.  We need to force that to happen, possibly just by [intersecting](https://www.typescriptlang.org/docs/handbook/2/objects.html#intersection-types) with `T`:

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
  (T extends { value: infer V } ?
    Omit<T, "value"> & { value: Exclude<V, number> } : never) & T;

declare function numberGuard<T extends { value: unknown; }>(
  v: T): v is ExcludeNumberFromValue<T> // okay
```

but while conceptually that's right, intersection tends to have some weird effects:

```ts
/* type Y = {
  x: Date;
  value: string;
} | {
  x: Date;
  value: never;
  y: number;
} | {
  y: number;
  value: boolean;
}*/
```

That `{value: never}` is weird.  In this case I think we can use [`Extract`](https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union) instead:

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
  Extract<T extends { value: infer V } ?
    Omit<T, "value"> & { value: Exclude<V, number> } : never, T>;
```

That works, and `Y` looks correct... *if* you use the `Id<T>` utility type. We could leave it there, but then your output type might be a bit unpleasant to look at:

```ts
const arr = polyRelationshipMany(e.items);
/* const arr: ((Omit<{
    relationTo: "type-a";
    value: number | TypeA;
}, "value"> & {
    value: TypeA;
}) | (Omit<{
    relationTo: "type-b";
    value: number | TypeB;
}, "value"> & {
    value: TypeB;
}) | (Omit<...> & {
    ...;
}))[];
```

So let's build in that `Id<T>` utility type in there.  It looks like

```ts
type Id<T> = T extends infer U ? { [K in keyof U]: U[K] } : never;
```

which is really just splitting a union into members, doing an "identity" [mapped type](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html) over each member, and pushing the union together again.  We want to put that inline, and it should be *inside* the `Extract` since we want the whole thing to be assignable to `T`.  Now we have:

```ts
type ExcludeNumberFromValue<T extends { value: unknown }> =
  Extract<(T extends { value: infer V } ?
    Omit<T, "value"> & { value: Exclude<V, number> } : never
  ) extends infer U ? { [K in keyof U]: U[K] } : never, T>;
```

and 

```ts
const arr = polyRelationshipMany(e.items);
/* const arr: (
  { relationTo: "type-a"; value: TypeA; } |
  { relationTo: "type-b"; value: TypeB; } | 
  { relationTo: "type-c"; value: TypeC; }
)[] */

arr.map(x => {
  switch (x.relationTo) {
    case "type-a": return x.value.a; // okay
    case "type-b": return x.value.b; // okay
    case "type-c": return x.value.c; // okay
  }
})
```

That looks like what we wanted.  The type of `arr`'s elements looks like `Example` with all the `number` parts of `value` filtered out without affecting the rest of the shape.  And thus we can treat each element like the discriminated union it is.

[Playground link to code](https://www.typescriptlang.org/play/?target=11#code/JYOwLgpgTgZghgYwgAgCoE8AOECCyDeAUMsnAFzICMhAvoaJLIihtgEIHHIBGFATLXrho8JGiwQAwpxIIKAZkFgJyAKIAPOAFtMAGxQBeGcmCQtAZzJcAPsgAUXEraIlXyKBF1wwwAPYhUXwoAcmVsAFo4YIBuR1cANzhdAFcIChBkrW5oZFtWXFi3OjdnOJIPLx9-QJCwiHDuGLLkRJS05AysnLyJNkLXYtdSt3LPbz8AoORQiXCEJpGWpNT0zOyoXPFsSX6SOi4ASgBtAF1BABMIBC8PZAR-czBkdo1tPQhCT5BtCHNMZmQAHkoMAAOagJKcLgwZIgBBVEAdNbQADiyTgUHOAB5UAA+BxuUwQLQUFwjCrjapTR4gkCggA0zVaKyRXQ2eTigwOFCJWhM5mMo0qExqyBpoAZcWZ7VQXBogvcEDAySgiLqvhgJjMADppcgAIQGIwAIk662Ncs+JBhcIRyEwvl06AASmMEeYABbATAAWTgIHQOPxcV5lgVQspkwo4rpjMW0tWbM2sqKp0OCo8ytVWuJ5m1MGAukYdjNqPRmIO-X2JAg2tD2q0cEwdnUyAMuIVwE1dnVmvUuuWhiNyFNyKgxoOiqziIADLsxQB3UwID32fsUhGBSdktwIODmFDGuqRY0UTMqxH96XauDz2T7w-H7inqcX5BXwfa7h3u4PkfHhAX3PbMPzabUEHnQYaErK07geJ4MQ2IwHSdV1hQeL1fX9QN8jwHp2GTCRJHxWtQxgkgAHoACo4JAR5SCgKBSTiDcRWpMBaVBecEy2XBCII-CpF2GhTmQKiKNgxCGybFs2w7HdF2XVcW21ViqW3Zo9wPf9ZjgIClTfUDUhvaJkAoijnkY3woH1TS-yPWZnzPAyQIHMDvzMizoCgazbMWLTH1mQDnOnd83OMiDPMsnybM5OUYP2OpkAASWxPE2zQZ51EgEBzgFUAYByABVZAAH4CGQI4AGkTERABrCB0A1ZAipOCgiuqk5kHldIIHiaBYm+LRfn+MQcDAMxMDAGcoRIJKNGuZJLgAOTHAAxHytAANUHHEspyvKKp42E6pAXwF0RGgOwMOJAS0UwcXpEdpWNDsADIjsHCgFpSS4cSOY0XpOJ7SygDsaFiT4uCSgANDKfqWiBVrZDbfG23b8HfCgABFvAgJ6eJjUFNlB7rcS4czkFhjK7oezH1BxvGCa+sUOIlEmx26p7AcHV7kA+zHCbZuluuhlRUrhoxUqxGHycoiykoljL6cZyBTKFzjTOrKmVAATQy6WEZW9bNp2tosRV5BccgZm2mjYXidsUn5WcZB0ETdZbZZbhfEdCB-Q5pMrrlqKkv1oxBZZonNh9v3-S1qH9iGkaAXGyawEoWadewNR1EW42UdN3bUH2iBcoFSO7eQE6zousm2ziEuIGysvDpr87EVK277rAR7nt597Pqro2IH+nm2mNYHWXWcG4l6-qoEhsWc8l3P86Rk20bN1ILaxq2maWKvo6dzmrophWVFX2ne8t638cPllj+nnIaG5l7B8rx+HdFubxfOVfpay3PtnFASsI57zvurKODsE7LxQOHFK2IR7I3WKjdG5tb4Hw1uzE+QdNiY3ds-KAXt2ix30AHXB6wyYh2ogqBm+81ZSmgZrOU+C4iENBtxFmZD-YgCrGJCSnwk4-D+KnCaxIpp8CzvNPOv0N6Fy3sXUu5ch4snbnXK6Dc3BNxbioz+7QCo5C2t1Mq3cHqoDfgPfmqiXiyMRliLaIMxzg2QPPAaUNf4r3hnYguqCi4YIgVg5hOCiFk2AdTIw19d70LviQ+2nFA5UNfv3CeH8H7tGjoMRW-8DbYiAfLEBiDV6YLVuk+JEpYGePgbkrEyDN7oJ3iU++2CRaUJfvgt2HtoBxJ4L7chiI2kbGDhTGiCkYl4y4UfGBrDhiuA4WOSZ3s+m8P4eJWClxrgYhQDaeEEwiFogxGlZRh19EUHUXwsmBIEgUFQNyFo-I15yJQdANB29R7pUpt5XywC8BJUwB4c4wA9yQGCAKJKWhkj0WyKQcw5gwTfG4PoKmvgtQCn+FAH4jBQUgO1K4QQXwRGjRQGnCRYB5DSJUHUhRDT3nHIrmU6uIBTod3rjdNwdgdEHXpTxQxGxjHyi7osKJFiUmpD5gLBlI8HFOLZC4txUBJwfVQJDEgGybjbNhLs-w+zyxHOblymxZymW1wuVdK5JB4g3LufEB5VK-GKPNh8iyvg6pwHQLBCJjzEbPKgK8jGgSbYMqfs7GhF8vGRJ7tE1WzTgmtNCck8eYq0ktOJlkv+AC8mhsKWAiq4zSkpsqYUhBhsfHyPtTSqNDCY1TISYMkxrt5lsh6Twih8bcQ0JomHZWcQ80QEWRk6ZAw2FuF7f2+V85G3rGEsOuZXTF5MKri2vhco1n7EIMnURY1xE6DAAAFgpTnO1Lz-E705a3blLNzmsriBoDiiBe5nr0Qy3lyB+WmKFZGkViaIDisNV6v6jiiFyo6H1bpaBcQquQGqrZyAdl2lBgczEe19Xnr-ecrWwY3CWrQNa21pafV+sdR2SmLq3Uesvt49eBGT2jyaT04Np8s2eqiXRoN3860JvftY05rMElppzjmwBTH03KwDX2tjmtRZwOQMWpB+H6lvMrbEiTITnYdMnWBniS7Ekv3bSMuh0ax28YqTMhUGn51Ye4cs+OrC1nQk1XaFCLo3QTE9N6P0AZkO6JOQyq9ZqQxmDDKgNMJA7lHt9TRoMokFLAURPWAsRZoAljHIh845Ef60XoohDKTm0KRjc1hAMdhSKBfS7Q+4dEEKMQoHYOwLHmhqSjDpCIekfw8VJoJHAVZLGpO40yFmuEqyTlsHVyNCkIybimA5CIz42ssw67xPocoetJr6-GAbvQhubFGw9bUe2P7ND29qIbxwTjRDs-iqGMiqMKaUShp9x1jUss0Wy3Od74RYg5XSv9L632CtcMKlbP7k0sylYB0GwGQCgagOme7h0X0lXKpjaqtVkANSapqVq7VOomPlU9PEkM4N7IQ7qrzBqeN+fxNh25VgLV4Zu9SxT6UySxZAc1eI4UUCGhNKDC0+wifatyy5jC7nsJk9QxTp7GjMM5gsDctMYX5OM+LriaLXBWfxcLMWEnhyEpQxK7maSzZWztmMF2ewvYwp6iNDzscE5XzZjnFwcwS4wArjXKpYXkwNK7nssePSIVDKc5MnEAKzX6hOQd5eYP35Q9+6CvpUKRlawQUtNBT4FWsuMRy46Zz6E6KYQ8+gYrdZSuxHK-BBiTF7BcExo10U036itYZYN+ttfFT54b0+Y0UCq75D6PW5A7f69TYAj3lvREE6nYEYQKSjZjdyWMC7pSHuR8+9cGHxvJ5A+uWvLeKKpH3W++0lvyPrPk9flMiR11R+N-x4iMFKPVvPyRWv2RvYtADhAA)

**Comment by jsardev** [comment]
I've worked a few iterations with AI and we came up with something like this:

```
type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

type Populated<T, D extends number> = D extends 0
  ? T
  : T extends any[]
    ? T extends (infer Item)[]
      ? Populated<Item, Prev[D]>[]
      : never
    : T extends object
      ? {
          [K in keyof T]: T[K] extends (infer Item)[]
            ? Exclude<Item, number> extends never
              ? T[K]
              : Populated<Exclude<Item, number>, Prev[D]>[]
            : Exclude<T[K], number> extends never
              ? T[K] extends object
                ? Populated<T[K], Prev[D]>
                : T[K]
              : Populated<Exclude<T[K], number>, Prev[D]>
        }
      : T;
```
and now:
```
type User = {
  id: number;
  name: string;
  cart: number | Cart;
}

type Cart = {
  id: number;
  items: string[];
}

type Post = {
  id: number;
  author: number | User;
}

type Test1 = Populated<Post, 1>; // post.author: User, post.author.cart: number | Cart
type Test2 = Populated<Post, 2>; // post.author User, post.author.cart: Cart
```

It doesn't solve the problem globally, but you can create populated types a lot easier – maybe it helps someone 😃
